File: src-tauri/src/bin/roomai.rs
// roomai — a tiny, honest, offline CLI for a Arcelle file.
//
// It reuses the desktop app's library crate (`arcelle_lib`) so it opens
// rooms with the exact same SQLCipher scheme the app uses — no second code
// path to drift out of sync. Nothing here touches the network or a model; it
// only decrypts a local file and reads or copies out what is already inside.
//
// Contract section E. Subcommands: verify / info / recover / export.
// Secrets come from the environment ONLY (ROOMAI_PASSWORD / ROOMAI_RECOVERY).
// There is deliberately no --password / --code flag: on macOS every process
// running as you can read another's argv (`ps -ww`, /proc equivalents, any
// monitoring agent), so a room password typed on the command line is readable
// by anything on the machine — and lands in shell history besides. The flags
// used to exist as an "escape hatch"; the environment is the escape hatch.
// Exit codes: 0 ok, 1 runtime error, 2 usage.

use std::collections::HashSet;
use std::path::Path;

use arcelle_lib::db;
use rusqlite::Connection;

const USAGE: &str = "\
roomai — offline tools for a Arcelle file

Usage:
  roomai verify  <path>           Check the file decrypts and is a valid room.
  roomai info    <path>           Verify, then dump the room's meta keys.
  roomai recover <path>           Open using a recovery code instead of a password.
  roomai export  <path> <outdir>  Write every stored file back out to a folder.

Secrets (environment only — never on the command line, where any other
program running as you can read them):
  ROOMAI_PASSWORD   password for verify / info / export
  ROOMAI_RECOVERY   recovery code for recover

Exit codes: 0 ok, 1 error, 2 usage.
";

/// A parsed command line. Splitting parse from run keeps argument handling
/// pure and unit-testable (no room, no filesystem).
#[derive(Debug, PartialEq, Eq)]
enum Command {
    Verify {
        path: String,
    },
    Info {
        path: String,
    },
    Recover {
        path: String,
    },
    Export {
        path: String,
        outdir: String,
    },
    /// No args, `help`, or `-h/--help`. Renders usage and exits 2.
    Help,
}

fn main() {
    // argv[0] is the program name; the parser works on everything after it.
    let args: Vec<String> = std::env::args().skip(1).collect();
    std::process::exit(run(args));
}

/// Top-level dispatch. Returns the process exit code and never panics on bad
/// input — every failure path maps to a clean stderr line and a non-zero code.
fn run(args: Vec<String>) -> i32 {
    let cmd = match parse(args) {
        Ok(c) => c,
        Err(msg) => {
            eprintln!("{msg}\n");
            eprint!("{USAGE}");
            return 2;
        }
    };

    let result = match cmd {
        Command::Help => {
            eprint!("{USAGE}");
            return 2;
        }
        Command::Verify { path } => do_verify(&path),
        Command::Info { path } => do_info(&path),
        Command::Recover { path } => do_recover(&path),
        Command::Export { path, outdir } => do_export(&path, &outdir),
    };

    match result {
        Ok(()) => 0,
        Err(e) => {
            eprintln!("Error: {e}");
            1
        }
    }
}

// ----------------------------------------------------------------- parsing

fn parse(mut args: Vec<String>) -> Result<Command, String> {
    if args.is_empty() {
        return Ok(Command::Help);
    }
    let cmd = args.remove(0);
    match cmd.as_str() {
        "help" | "-h" | "--help" => Ok(Command::Help),
        "verify" | "info" => {
            reject_secret_flags(&args)?;
            let path = one_positional(args, &cmd, "<path>")?;
            Ok(if cmd == "verify" {
                Command::Verify { path }
            } else {
                Command::Info { path }
            })
        }
        "recover" => {
            reject_secret_flags(&args)?;
            let path = one_positional(args, &cmd, "<path>")?;
            Ok(Command::Recover { path })
        }
        "export" => {
            reject_secret_flags(&args)?;
            let (path, outdir) = two_positionals(args, &cmd)?;
            Ok(Command::Export { path, outdir })
        }
        other => Err(format!("Unknown command: {other}")),
    }
}

/// The two flags that used to carry a secret on the command line. Named
/// explicitly rather than falling through to "Unknown flag" so anyone with an
/// old script is told WHY it stopped working and what to do instead.
fn reject_secret_flags(args: &[String]) -> Result<(), String> {
    for a in args {
        if a == "--password" || a == "--code" {
            return Err(format!(
                "`{a}` is no longer accepted: any program running as you can read \
                 another process's arguments, so a secret there is not private. \
                 Set ROOMAI_PASSWORD / ROOMAI_RECOVERY in the environment instead."
            ));
        }
    }
    Ok(())
}

/// Reject any leftover token that looks like a flag — catches typos such as
/// `--passwrod` before they get silently treated as a path.
fn reject_unknown_flags(args: &[String]) -> Result<(), String> {
    for a in args {
        if a.starts_with('-') && a != "-" {
            return Err(format!("Unknown flag: {a}"));
        }
    }
    Ok(())
}

fn one_positional(args: Vec<String>, cmd: &str, what: &str) -> Result<String, String> {
    reject_unknown_flags(&args)?;
    match args.len() {
        0 => Err(format!("`{cmd}` needs a {what}.")),
        1 => Ok(args.into_iter().next().expect("len checked")),
        _ => Err(format!("`{cmd}` takes exactly one {what}.")),
    }
}

fn two_positionals(args: Vec<String>, cmd: &str) -> Result<(String, String), String> {
    reject_unknown_flags(&args)?;
    if args.len() != 2 {
        return Err(format!("`{cmd}` needs <path> and <outdir>."));
    }
    let mut it = args.into_iter();
    Ok((it.next().unwrap(), it.next().unwrap()))
}

// --------------------------------------------------------------- secrets

fn resolve_password() -> Result<String, String> {
    match std::env::var("ROOMAI_PASSWORD") {
        Ok(v) if !v.is_empty() => Ok(v),
        _ => Err("No password. Set ROOMAI_PASSWORD (never pass it as an argument — \
                  other programs on this Mac can read a process's arguments)."
            .into()),
    }
}

fn resolve_code() -> Result<String, String> {
    match std::env::var("ROOMAI_RECOVERY") {
        Ok(v) if !v.is_empty() => Ok(v),
        _ => Err("No recovery code. Set ROOMAI_RECOVERY (never pass it as an argument — \
                  other programs on this Mac can read a process's arguments)."
            .into()),
    }
}

// --------------------------------------------------------------- commands

/// CHECK, don't touch. `db::open_room` always runs the schema migration and at
/// least one write, so "verify" used to fail outright on a copy sitting on a
/// read-only volume — and on a room not opened since the last schema change it
/// rebuilt the search index while claiming only to be looking.
fn do_verify(path: &str) -> Result<(), String> {
    let password = resolve_password()?;
    let conn = db::open_room_readonly(path, &password)?;
    print_ok(&conn);
    Ok(())
}

fn do_info(path: &str) -> Result<(), String> {
    let password = resolve_password()?;
    let conn = db::open_room_readonly(path, &password)?;
    print_ok(&conn);
    print_meta_dump(&conn);
    Ok(())
}

fn do_recover(path: &str) -> Result<(), String> {
    let code = resolve_code()?;
    let conn = db::open_with_recovery(path, &code)?;
    print_ok(&conn);
    Ok(())
}

fn do_export(path: &str, outdir: &str) -> Result<(), String> {
    let password = resolve_password()?;
    let conn = db::open_room(path, &password)?;

    std::fs::create_dir_all(outdir)
        .map_err(|e| format!("Could not create {outdir}: {e}"))?;

    let mut stmt = conn
        // Trash: an export is "everything in this room", and a deleted file is
        // not in the room. Exporting it would also be the one way trashed bytes
        // could reach the plain filesystem — the whole point of keeping them
        // inside the room is that they never do.
        .prepare(
            "SELECT name, original_bytes FROM files WHERE trashed_at IS NULL \
             ORDER BY created_at, rowid",
        )
        .map_err(|e| e.to_string())?;
    let rows = stmt
        .query_map([], |r| {
            let name: String = r.get(0)?;
            let bytes: Option<Vec<u8>> = r.get(1)?;
            Ok((name, bytes))
        })
        .map_err(|e| e.to_string())?;

    let mut used: HashSet<String> = HashSet::new();
    // Never clobber what's already in the output folder: seed the dedupe
    // set with the names that exist there so unique_name() steps past them.
    // Keys are lowercased because APFS is case-insensitive by default, so
    // "Report.pdf" would overwrite an existing "report.pdf".
    let entries = std::fs::read_dir(outdir).map_err(|e| format!("Could not read {outdir}: {e}"))?;
    for entry in entries {
        let entry = entry.map_err(|e| e.to_string())?;
        used.insert(entry.file_name().to_string_lossy().to_lowercase());
    }
    let mut written = 0usize;
    let mut skipped = 0usize;

    for row in rows {
        let (name, bytes) = row.map_err(|e| e.to_string())?;
        // Rows can carry no original bytes (e.g. app-generated notes). Nothing
        // to write out — count it and move on rather than emit an empty file.
        let Some(bytes) = bytes else {
            skipped += 1;
            continue;
        };
        let filename = unique_name(&mut used, &sanitize(&name));
        let dest = Path::new(outdir).join(&filename);
        std::fs::write(&dest, &bytes)
            .map_err(|e| format!("Could not write {}: {e}", dest.display()))?;
        written += 1;
    }

    println!("Wrote {written} file(s) to {outdir}.");
    if skipped > 0 {
        println!("Skipped {skipped} row(s) with no stored bytes.");
    }
    Ok(())
}

// ---------------------------------------------------------------- helpers

/// Success banner shared by verify / info / recover.
fn print_ok(conn: &Connection) {
    println!("OK: decrypts and is a valid Arcelle file.");
    println!("  format:         {}", meta_or(conn, "format"));
    println!("  format_version: {}", meta_or(conn, "format_version"));
    println!("  files:          {}", count(conn, "files"));
    println!("  chats:          {}", count(conn, "chats"));
}

/// The extra `info` section: the interesting meta keys, shown even when unset
/// so it's clear which ones a room does and doesn't carry.
fn print_meta_dump(conn: &Connection) {
    println!("\nMeta:");
    for key in ["format", "format_version", "name", "embed_model", "embed_dim"] {
        match db::get_meta(conn, key) {
            Some(v) => println!("  {key}: {v}"),
            None => println!("  {key}: (not set)"),
        }
    }
}

fn meta_or(conn: &Connection, key: &str) -> String {
    db::get_meta(conn, key).unwrap_or_else(|| "(unset)".into())
}

/// Count rows in a table. The table name is always a hard-coded literal from
/// this file, never user input, so the format! is not an injection surface.
fn count(conn: &Connection, table: &str) -> i64 {
    conn.query_row(&format!("SELECT count(*) FROM {table}"), [], |r| r.get(0))
        .unwrap_or(0)
}

/// Minimal filename hardening for export: keep only the final path component
/// and neutralise separators / NUL, so a stored name like `../../etc/passwd`
/// can never write outside the chosen output directory.
fn sanitize(name: &str) -> String {
    let base = name.rsplit(['/', '\\']).next().unwrap_or(name);
    let cleaned: String = base
        .chars()
        .map(|c| match c {
            '/' | '\\' | '\0' => '_',
            _ => c,
        })
        .collect();
    let trimmed = cleaned.trim();
    if trimmed.is_empty() || trimmed == "." || trimmed == ".." {
        "unnamed".to_string()
    } else {
        trimmed.to_string()
    }
}

/// Make `base` unique against `used` — seeded with the destination's
/// pre-existing filenames and every name handed out earlier in this export
/// run — so nothing gets silently clobbered. Appends ` (2)`, ` (3)`, … before
/// the extension. `used` holds lowercased keys (and candidates are tested
/// lowercased) because APFS is case-insensitive by default; the returned
/// name keeps the original case.
fn unique_name(used: &mut HashSet<String>, base: &str) -> String {
    if used.insert(base.to_lowercase()) {
        return base.to_string();
    }
    let (stem, ext) = match base.rsplit_once('.') {
        Some((s, e)) if !s.is_empty() => (s.to_string(), format!(".{e}")),
        _ => (base.to_string(), String::new()),
    };
    let mut n = 2;
    loop {
        let candidate = format!("{stem} ({n}){ext}");
        if used.insert(candidate.to_lowercase()) {
            return candidate;
        }
        n += 1;
    }
}

// ------------------------------------------------------------------ tests

#[cfg(test)]
mod tests {
    use super::*;

    fn argv(parts: &[&str]) -> Vec<String> {
        parts.iter().map(|s| s.to_string()).collect()
    }

    #[test]
    fn no_args_is_help() {
        assert_eq!(parse(argv(&[])).unwrap(), Command::Help);
    }

    #[test]
    fn help_word_and_flags_are_help() {
        assert_eq!(parse(argv(&["help"])).unwrap(), Command::Help);
        assert_eq!(parse(argv(&["-h"])).unwrap(), Command::Help);
        assert_eq!(parse(argv(&["--help"])).unwrap(), Command::Help);
    }

    #[test]
    fn unknown_command_is_rejected() {
        assert!(parse(argv(&["frobnicate", "room.room"])).is_err());
    }

    #[test]
    fn verify_requires_a_path() {
        assert!(parse(argv(&["verify"])).is_err());
    }

    /// A room password on the command line is readable by every other process
    /// running as you (`ps -ww`), and lands in shell history. The flag that
    /// accepted one is refused BY NAME in every position, with a message that
    /// says where to put the secret instead — an old script must not fail with
    /// a bare "Unknown flag" and must certainly not silently keep working.
    #[test]
    fn a_secret_on_the_command_line_is_refused_by_name() {
        for argv_ in [
            argv(&["verify", "room.room", "--password", "hunter2"]),
            argv(&["verify", "--password", "hunter2", "room.room"]),
            argv(&["info", "room.room", "--password", "hunter2"]),
            argv(&["export", "room.room", "out", "--password", "hunter2"]),
            argv(&["recover", "room.room", "--code", "AAAA-BBBB"]),
        ] {
            let err = parse(argv_).expect_err("a secret argument must be refused");
            assert!(err.contains("ROOMAI_PASSWORD") || err.contains("ROOMAI_RECOVERY"), "{err}");
        }
    }

    #[test]
    fn verify_takes_only_a_path() {
        assert_eq!(
            parse(argv(&["verify", "room.room"])).unwrap(),
            Command::Verify { path: "room.room".into() }
        );
    }

    #[test]
    fn info_parses() {
        assert_eq!(
            parse(argv(&["info", "room.room"])).unwrap(),
            Command::Info { path: "room.room".into() }
        );
    }

    #[test]
    fn recover_takes_only_a_path() {
        assert_eq!(
            parse(argv(&["recover", "room.room"])).unwrap(),
            Command::Recover { path: "room.room".into() }
        );
    }

    #[test]
    fn export_needs_two_positionals() {
        assert!(parse(argv(&["export", "room.room"])).is_err());
        assert!(parse(argv(&["export", "room.room", "out", "extra"])).is_err());
        assert_eq!(
            parse(argv(&["export", "room.room", "out"])).unwrap(),
            Command::Export {
                path: "room.room".into(),
                outdir: "out".into(),
            }
        );
    }

    #[test]
    fn unknown_flag_is_rejected() {
        assert!(parse(argv(&["verify", "room.room", "--nope"])).is_err());
        assert!(parse(argv(&["verify", "--passwrod", "x", "room.room"])).is_err());
    }

    #[test]
    fn a_bare_secret_flag_errors_too() {
        // No value to take, so this must not fall through to "<path>".
        assert!(parse(argv(&["verify", "room.room", "--password"])).is_err());
    }

    #[test]
    fn sanitize_strips_path_traversal() {
        assert_eq!(sanitize("../../etc/passwd"), "passwd");
        assert_eq!(sanitize("a/b/c.txt"), "c.txt");
        assert_eq!(sanitize("plain.pdf"), "plain.pdf");
        assert_eq!(sanitize(""), "unnamed");
        assert_eq!(sanitize("   "), "unnamed");
        assert_eq!(sanitize(".."), "unnamed");
    }

    #[test]
    fn unique_name_disambiguates_collisions() {
        let mut used = HashSet::new();
        assert_eq!(unique_name(&mut used, "a.txt"), "a.txt");
        assert_eq!(unique_name(&mut used, "a.txt"), "a (2).txt");
        assert_eq!(unique_name(&mut used, "a.txt"), "a (3).txt");
        assert_eq!(unique_name(&mut used, "noext"), "noext");
        assert_eq!(unique_name(&mut used, "noext"), "noext (2)");
    }

    #[test]
    fn unique_name_steps_past_preexisting_names() {
        // Names seeded from the destination directory must not be reused.
        let mut used = HashSet::new();
        used.insert("report.pdf".to_string());
        assert_eq!(unique_name(&mut used, "report.pdf"), "report (2).pdf");
    }

    #[test]
    fn unique_name_is_case_insensitive() {
        // APFS is case-insensitive by default: "Report.pdf" would clobber a
        // pre-existing "report.pdf". Keys are compared lowercased, but the
        // returned name keeps the caller's case.
        let mut used = HashSet::new();
        used.insert("report.pdf".to_string()); // seeded (already lowercased)
        assert_eq!(unique_name(&mut used, "Report.pdf"), "Report (2).pdf");
        assert_eq!(unique_name(&mut used, "REPORT.PDF"), "REPORT (3).PDF");
        assert_eq!(unique_name(&mut used, "Fresh.pdf"), "Fresh.pdf");
        assert_eq!(unique_name(&mut used, "fresh.pdf"), "fresh (2).pdf");
    }
}

File: apps/desktop/cli/roomai.ts
/**
 * roomai — a tiny, honest, offline CLI for a Arcelle file.
 *
 * Ported from `src-tauri/src/bin/roomai.rs` (the whole file, including its
 * module doc comment and every `#[test]`).
 *
 * It reuses the app's own db-host modules (`open.ts` / `recovery.ts` /
 * `meta.ts`) so it opens rooms with the exact same SQLCipher scheme the app
 * uses — no second code path to drift out of sync. Nothing here touches the
 * network or a model; it only decrypts a local file and reads or copies out
 * what is already inside.
 *
 * Subcommands: verify / info / recover / export. Secrets come from the
 * environment ONLY (ROOMAI_PASSWORD / ROOMAI_RECOVERY). There is
 * deliberately no --password / --code flag: on macOS every process running
 * as you can read another's argv (`ps -ww`, /proc equivalents, any
 * monitoring agent), so a room password typed on the command line is
 * readable by anything on the machine — and lands in shell history besides.
 * The flags used to exist as an "escape hatch"; the environment is the
 * escape hatch.
 *
 * Splitting `parseArgs` (pure) from the `do*` functions (real filesystem/DB
 * work) keeps argument handling pure and unit-testable (no room, no
 * filesystem) — mirrored from the Rust module's own `parse`/`run` split,
 * called out in its doc comment.
 *
 * Exit codes: 0 ok, 1 runtime error, 2 usage.
 *
 * NOTE (deviation from the Rust source, deliberate): the Rust `do_verify` /
 * `do_info` / `do_recover` / `do_export` each call `println!` themselves as
 * they go. Here the `do*` functions instead BUILD and RETURN the output as a
 * string, and only `main()` prints it, so the formatting logic is testable
 * without capturing real stdout. The exact text (including the blank line
 * Rust's `print_meta_dump`'s leading `println!("\nMeta:")` puts between the
 * OK banner and the "Meta:" section) is preserved.
 *
 * NOTE (deviation, deliberate): `main()` here does not call `process.exit()`
 * itself — it returns the exit code so it stays testable. A real bin
 * entrypoint wires that up (see the bottom of this file).
 */

import * as fsp from "node:fs/promises";
import * as nodePath from "node:path";
import type Database from "better-sqlite3-multiple-ciphers";
import { openRoom, openRoomReadonly } from "../src/main/db-host/open.js";
import { openWithRecovery } from "../src/main/db-host/recovery.js";
import { getMeta } from "../src/main/db-host/meta.js";

// ------------------------------------------------------------------ usage

export const USAGE =
  "roomai — offline tools for a Arcelle file\n" +
  "\n" +
  "Usage:\n" +
  "  roomai verify  <path>           Check the file decrypts and is a valid room.\n" +
  "  roomai info    <path>           Verify, then dump the room's meta keys.\n" +
  "  roomai recover <path>           Open using a recovery code instead of a password.\n" +
  "  roomai export  <path> <outdir>  Write every stored file back out to a folder.\n" +
  "\n" +
  "Secrets (environment only — never on the command line, where any other\n" +
  "program running as you can read them):\n" +
  "  ROOMAI_PASSWORD   password for verify / info / export\n" +
  "  ROOMAI_RECOVERY   recovery code for recover\n" +
  "\n" +
  "Exit codes: 0 ok, 1 error, 2 usage.\n";

// ----------------------------------------------------------------- parsing

/**
 * A parsed command line. Splitting parse from run keeps argument handling
 * pure and unit-testable (no room, no filesystem).
 */
export type Command =
  | { kind: "verify"; path: string }
  | { kind: "info"; path: string }
  | { kind: "recover"; path: string }
  | { kind: "export"; path: string; outdir: string }
  /** No args, `help`, or `-h/--help`. Renders usage and exits 2. */
  | { kind: "help" };

/**
 * The two flags that used to carry a secret on the command line. Named
 * explicitly rather than falling through to "Unknown flag" so anyone with an
 * old script is told WHY it stopped working and what to do instead.
 */
function rejectSecretFlags(args: string[]): void {
  for (const a of args) {
    if (a === "--password" || a === "--code") {
      throw new Error(
        `\`${a}\` is no longer accepted: any program running as you can read another process's arguments, so a secret there is not private. Set ROOMAI_PASSWORD / ROOMAI_RECOVERY in the environment instead.`
      );
    }
  }
}

/**
 * Reject any leftover token that looks like a flag — catches typos such as
 * `--passwrod` before they get silently treated as a path.
 */
function rejectUnknownFlags(args: string[]): void {
  for (const a of args) {
    if (a.startsWith("-") && a !== "-") {
      throw new Error(`Unknown flag: ${a}`);
    }
  }
}

function onePositional(args: string[], cmd: string, what: string): string {
  rejectUnknownFlags(args);
  if (args.length === 0) {
    throw new Error(`\`${cmd}\` needs a ${what}.`);
  }
  if (args.length === 1) {
    return args[0]!;
  }
  throw new Error(`\`${cmd}\` takes exactly one ${what}.`);
}

function twoPositionals(args: string[], cmd: string): [string, string] {
  rejectUnknownFlags(args);
  if (args.length !== 2) {
    throw new Error(`\`${cmd}\` needs <path> and <outdir>.`);
  }
  return [args[0]!, args[1]!];
}

/**
 * Parse `args` (already stripped of the program name — i.e. what
 * `process.argv.slice(2)` gives you) into a `Command`. Throws `Error(message)`
 * on any parse failure. No args, `help`, or `-h`/`--help` is
 * `{ kind: "help" }`, not a throw.
 */
export function parseArgs(args: string[]): Command {
  if (args.length === 0) {
    return { kind: "help" };
  }
  const cmd = args[0]!;
  const rest = args.slice(1);
  switch (cmd) {
    case "help":
    case "-h":
    case "--help":
      return { kind: "help" };
    case "verify":
    case "info": {
      rejectSecretFlags(rest);
      const path = onePositional(rest, cmd, "<path>");
      return cmd === "verify" ? { kind: "verify", path } : { kind: "info", path };
    }
    case "recover": {
      rejectSecretFlags(rest);
      const path = onePositional(rest, cmd, "<path>");
      return { kind: "recover", path };
    }
    case "export": {
      rejectSecretFlags(rest);
      const [path, outdir] = twoPositionals(rest, cmd);
      return { kind: "export", path, outdir };
    }
    default:
      throw new Error(`Unknown command: ${cmd}`);
  }
}

// --------------------------------------------------------------- secrets

/** Throws if `ROOMAI_PASSWORD` is unset/empty. */
export async function resolvePassword(): Promise<string> {
  const v = process.env.ROOMAI_PASSWORD;
  if (v !== undefined && v !== "") {
    return v;
  }
  throw new Error(
    "No password. Set ROOMAI_PASSWORD (never pass it as an argument — other programs on this Mac can read a process's arguments)."
  );
}

/** Throws if `ROOMAI_RECOVERY` is unset/empty. */
export async function resolveCode(): Promise<string> {
  const v = process.env.ROOMAI_RECOVERY;
  if (v !== undefined && v !== "") {
    return v;
  }
  throw new Error(
    "No recovery code. Set ROOMAI_RECOVERY (never pass it as an argument — other programs on this Mac can read a process's arguments)."
  );
}

// ---------------------------------------------------------------- helpers

/** The value, or "(unset)" when the key is missing. */
function metaOr(db: Database.Database, key: string): string {
  return getMeta(db, key) ?? "(unset)";
}

/**
 * Count rows in a table. The table name is always a hard-coded literal from
 * this file, never user input, so string interpolation is not an injection
 * surface.
 */
function countRows(db: Database.Database, table: string): number {
  try {
    const row = db.prepare(`SELECT count(*) AS c FROM ${table}`).get() as
      | { c: number }
      | undefined;
    return row?.c ?? 0;
  } catch {
    return 0;
  }
}

/** Success banner shared by verify / info / recover. */
function formatOk(db: Database.Database): string {
  return [
    "OK: decrypts and is a valid Arcelle file.",
    `  format:         ${metaOr(db, "format")}`,
    `  format_version: ${metaOr(db, "format_version")}`,
    `  files:          ${countRows(db, "files")}`,
    `  chats:          ${countRows(db, "chats")}`,
  ].join("\n");
}

/**
 * The extra `info` section: the interesting meta keys, shown even when unset
 * so it's clear which ones a room does and doesn't carry.
 *
 * Starts with a blank line then "Meta:" — mirroring Rust's own
 * `println!("\nMeta:")`, which puts a blank line between the OK banner
 * `print_ok` already printed and this section.
 */
function formatMetaDump(db: Database.Database): string {
  const lines: string[] = ["", "Meta:"];
  for (const key of ["format", "format_version", "name", "embed_model", "embed_dim"]) {
    const v = getMeta(db, key);
    lines.push(v !== null ? `  ${key}: ${v}` : `  ${key}: (not set)`);
  }
  return lines.join("\n");
}

/** Close a db handle, swallowing an already-closed/double-close error — this
 * runs in a `finally`, so a close failure must never mask whatever error (if
 * any) is already propagating out of the try block. Mirrors the same
 * defensive pattern `open.ts`'s own error paths use around `db.close()`. */
function closeQuietly(db: Database.Database): void {
  try {
    db.close();
  } catch {
    // already closed / nothing to do
  }
}

function errMessage(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

/**
 * Minimal filename hardening for export: keep only the final path component
 * and neutralise separators / NUL, so a stored name like `../../etc/passwd`
 * can never write outside the chosen output directory.
 */
export function sanitize(name: string): string {
  const segments = name.split(/[/\\]/);
  const base = segments[segments.length - 1] ?? name;
  let cleaned = "";
  for (const c of base) {
    cleaned += c === "/" || c === "\\" || c === "\0" ? "_" : c;
  }
  const trimmed = cleaned.trim();
  if (trimmed === "" || trimmed === "." || trimmed === "..") {
    return "unnamed";
  }
  return trimmed;
}

/**
 * Make `base` unique against `used` — seeded with the destination's
 * pre-existing filenames and every name handed out earlier in this export
 * run — so nothing gets silently clobbered. Appends ` (2)`, ` (3)`, … before
 * the extension. `used` holds LOWERCASED keys (and candidates are tested
 * lowercased) because APFS is case-insensitive by default; the returned name
 * keeps the original case.
 */
export function uniqueName(used: Set<string>, base: string): string {
  const lowerBase = base.toLowerCase();
  if (!used.has(lowerBase)) {
    used.add(lowerBase);
    return base;
  }
  const dot = base.lastIndexOf(".");
  const stem = dot > 0 ? base.slice(0, dot) : base;
  const ext = dot > 0 ? `.${base.slice(dot + 1)}` : "";
  let n = 2;
  for (;;) {
    const candidate = `${stem} (${n})${ext}`;
    const lowerCandidate = candidate.toLowerCase();
    if (!used.has(lowerCandidate)) {
      used.add(lowerCandidate);
      return candidate;
    }
    n += 1;
  }
}

// --------------------------------------------------------------- commands

/**
 * CHECK, don't touch. `openRoom` always runs the schema migration and at
 * least one write, so "verify" used to fail outright on a copy sitting on a
 * read-only volume — and on a room not opened since the last schema change
 * it rebuilt the search index while claiming only to be looking. Uses
 * `openRoomReadonly` instead.
 */
export async function doVerify(path: string): Promise<string> {
  const password = await resolvePassword();
  const db = openRoomReadonly(path, password);
  try {
    return formatOk(db);
  } finally {
    closeQuietly(db);
  }
}

export async function doInfo(path: string): Promise<string> {
  const password = await resolvePassword();
  const db = openRoomReadonly(path, password);
  try {
    return formatOk(db) + "\n" + formatMetaDump(db);
  } finally {
    closeQuietly(db);
  }
}

export async function doRecover(path: string): Promise<string> {
  const code = await resolveCode();
  const db = await openWithRecovery(path, code);
  try {
    return formatOk(db);
  } finally {
    closeQuietly(db);
  }
}

export async function doExport(path: string, outdir: string): Promise<string> {
  const password = await resolvePassword();
  const db = openRoom(path, password);
  try {
    try {
      await fsp.mkdir(outdir, { recursive: true });
    } catch (err) {
      throw new Error(`Could not create ${outdir}: ${errMessage(err)}`);
    }

    // Trash: an export is "everything in this room", and a deleted file is
    // not in the room. Exporting it would also be the one way trashed bytes
    // could reach the plain filesystem — the whole point of keeping them
    // inside the room is that they never do.
    const rows = db
      .prepare(
        "SELECT name, original_bytes FROM files WHERE trashed_at IS NULL ORDER BY created_at, rowid"
      )
      .all() as { name: string; original_bytes: Buffer | null }[];

    // Never clobber what's already in the output folder: seed the dedupe
    // set with the names that exist there so uniqueName() steps past them.
    // Keys are lowercased because APFS is case-insensitive by default, so
    // "Report.pdf" would overwrite an existing "report.pdf".
    const used = new Set<string>();
    let entries: string[];
    try {
      entries = await fsp.readdir(outdir);
    } catch (err) {
      throw new Error(`Could not read ${outdir}: ${errMessage(err)}`);
    }
    for (const entry of entries) {
      used.add(entry.toLowerCase());
    }

    let written = 0;
    let skipped = 0;
    for (const row of rows) {
      // Rows can carry no original bytes (e.g. app-generated notes). Nothing
      // to write out — count it and move on rather than emit an empty file.
      if (row.original_bytes === null || row.original_bytes === undefined) {
        skipped += 1;
        continue;
      }
      const filename = uniqueName(used, sanitize(row.name));
      const dest = nodePath.join(outdir, filename);
      try {
        await fsp.writeFile(dest, row.original_bytes);
      } catch (err) {
        throw new Error(`Could not write ${dest}: ${errMessage(err)}`);
      }
      written += 1;
    }

    let summary = `Wrote ${written} file(s) to ${outdir}.`;
    if (skipped > 0) {
      summary += `\nSkipped ${skipped} row(s) with no stored bytes.`;
    }
    return summary;
  } finally {
    closeQuietly(db);
  }
}

// ------------------------------------------------------------------- main

/**
 * Top-level dispatch. Returns the process exit code and never throws on bad
 * input — every failure path maps to a clean stderr line and a non-zero
 * code. Does NOT call `process.exit()` itself, so it stays testable; a real
 * bin entrypoint wires that up (see the bottom of this file).
 */
export async function main(argv: string[]): Promise<number> {
  let cmd: Command;
  try {
    cmd = parseArgs(argv);
  } catch (err) {
    process.stderr.write(`${errMessage(err)}\n\n`);
    process.stderr.write(USAGE);
    return 2;
  }

  if (cmd.kind === "help") {
    process.stderr.write(USAGE);
    return 2;
  }

  try {
    let output: string;
    switch (cmd.kind) {
      case "verify":
        output = await doVerify(cmd.path);
        break;
      case "info":
        output = await doInfo(cmd.path);
        break;
      case "recover":
        output = await doRecover(cmd.path);
        break;
      case "export":
        output = await doExport(cmd.path, cmd.outdir);
        break;
    }
    console.log(output);
    return 0;
  } catch (err) {
    console.error(`Error: ${errMessage(err)}`);
    return 1;
  }
}

// A real bin entrypoint is intentionally NOT included in this file — wire up
//   main(process.argv.slice(2)).then((code) => process.exit(code));
// in a tiny separate script once this becomes the actual shipped CLI, so the
// exported surface above stays free of a top-level side effect.
